### ISSUE TITLE
incorrect parsing of special characters in subvolume path

### ISSUE THREAD

**Issue by asssaf** [digint/btrbk#288]
I've been using this external drive for a while for backups. I've recently moved some snapshots manually (not sure if it's related or not) and now I'm getting this error when starting btrbk (even with something like `list snapshots`).

```
ERROR: process died unexpectedly (btrbk v0.28.2)
Please contact the author: Axel Burri <axel@tty0.ch>

Stack Trace:
----------------------------------------
Died at /usr/sbin/btrbk line 2273.
 at /usr/sbin/btrbk line 294.
        main::__ANON__("Died at /usr/sbin/btrbk line 2273.\x{a}") called at /usr/sbin/btrbk line 2273
        main::btr_tree(HASH(0x558f14558078), 5, "/dev/mapper/extbtrfs", ARRAY(0x558f14557e08)) called at /usr/sbin/btrbk line 2593
        main::vinfo_init_root(HASH(0x558f14529558)) called at /usr/sbin/btrbk line 5790
```

**Comment by digint** [comment]
Thanks for reporting, I fixed a bug:

7d656540d77b30a7c31922277fcd6247bb95eff4: btrbk: fix btr_tree for subvolumes containing special (unsafe) characters

Probably you have a subvolume name with special characters like `(` or `$`. While this is perfectly legal, it is generally a bad idea.

Allowed characters in btrbk are: `[0-9] [a-z] [A-Z] and "._+-@"`

With the bugfix, btrbk will not die when reading the subvolume list, but will still error out on any action (e.g. btrfs send -p) on these subvolumes:

    ERROR: Unsafe command [...] (offending string: "/tmp/btrfs_pool/svol with spaces")

Could you please try the patch and report back?
```
cd /tmp/
wget https://raw.githubusercontent.com/digint/btrbk/master/btrbk
chown +x btrbk
sudo ./btrbk list snapshots
```

**Comment by asssaf** [comment]
Thanks for the quick response! 
With the bugfix it indeed doesn't crash, but I got no warnings.
I think the problem was caused by having a subvolume with a nested subvolume with the same name. After removing it the original version works too.

**Comment by digint** [comment]
> I think the problem was caused by having a subvolume with a nested subvolume with the same name. After removing it the original version works too.

This sounds bad and should not happen. If you can still reproduce it, please paste the relevant parts of:

```
# btrbk master, >=v0.29.0
/tmp/btrbk ls /
/tmp/btrbk ls / --format=rel
/tmp/btrbk ls / --format=long
```

**Comment by asssaf** [comment]
Sorry, I tried but can't reproduce it.

**Comment by digint** [comment]
I tested several weird combos of subvolume / subdirs, it all seems to work.

Changed title, added bug label

fixed in: 7d656540d77b30a7c31922277fcd6247bb95eff4: btrbk: fix btr_tree for subvolumes containing special (unsafe) characters

**Comment by digint** [comment]
included in [btrbk-0.29.0](/digint/btrbk/releases/tag/v0.29.0)
